packages/gatsby/src/redux/nodes.ts

```diff
@@ -152,32 +166,65 @@ export const addResolvedNodes = (
   return resolvedNodes
 }
 
+export const postIndexingMetaSetup = (filterCache: IFilterCache): void => {
+  // Create an ordered array of individual nodes, ordered (grouped) by the
+  // value to which the filter resolves. Nodes are not ordered per value.
+  // This way non-eq ops can simply slice the array to get a range.
+
+  const entries: Array<[FilterValue, Set<IGatsbyNode>]> = [
+    ...filterCache.byValue.entries(),
+  ]
+
+  // Sort all sets by its value, asc. Ignore/allow potential type casting.
+  entries.sort(([a], [b]) => (a <= b ? -1 : a > b ? 1 : 0))
+
+  const orderedNodes: Array<IGatsbyNode> = []
+  const orderedValues: Array<FilterValue> = []
+  const offsets: Map<FilterValue, [number, number]> = new Map()
+  entries.forEach(([v, bucket]) => {
+    // Record the first index containing node with as filter value v
+    offsets.set(v, [orderedNodes.length, bucket.size])
+    // We could do `arr.push(...bucket)` here but that's not safe with very
+    // large sets, so we use a regular loop
+    bucket.forEach(node => orderedNodes.push(node))
```

Comment:
As far as I remember the same node can go into multiple buckets for `elemMatch`. Can this cause issues here as the node may appear on this list twice?

Reply:
That's a very good catch.

This is not a problem because when we collect all the nodes, we put all the buckets together in a Set. At this point the nodes will be deduped.

Worst case the ordered array can be n^2 in size. That would be some kind of weird elemMatch case. I don't expect this to be the case in the real world...